src/cryptsetup/cryptsetup-token-systemd-tpm2/cryptsetup-token-systemd-tpm2.c

```diff
@@ -0,0 +1,223 @@
+ */
+_public_ void cryptsetup_token_dump(
+                struct crypt_device *cd /* is always LUKS2 context */,
+                const char *json /* validated 'systemd-tpm2' token if cryptsetup_token_validate is defined */) {
+
+        bool first;
+        uint32_t i, pcr, pcr_mask;
+        _cleanup_free_ char* base64_blob = NULL, *hex_policy_hash = NULL;
+
+        if (parse_luks2_tpm2_data(json, UINT32_MAX, &pcr_mask, &base64_blob, &hex_policy_hash))
+                return;
+
+        crypt_log_std(cd, "\ttpm2-pcrs:  ");
+        for (i = 0, first = true; i < TPM2_PCRS_MAX; i++) {
+                pcr = pcr_mask & (UINT32_C(1) << i);
+                if (pcr) {
+                        crypt_log_std(cd, "%s%" PRIu32, first ? "" : ", ", pcr);
+                        first = false;
+                }
+        }
+
+        /* well, it's stored in plaintext json metadata anyway... */
+        crypt_log_std(cd, "\n\ttmp2-blob:  %s\n", base64_blob);
+
+        crypt_log_std(cd, "\ttmp2-policy-hash: %s\n", hex_policy_hash);
+}
+
+/*
+ * Note:
+ *   If plugin is available in library path, it's called in before following libcryptsetup calls:
+ *
+ *   crypt_token_json_set, crypt_dump, any crypt_activate_by_token_* flavour
+ */
+_public_ int cryptsetup_token_validate(
+                struct crypt_device *cd, /* is always LUKS2 context */
+                const char *json /* contains valid 'type' and 'keyslots' fields. 'type' is 'systemd-tpm2' */) {
+
+        int r;
+        JsonVariant *w, *e;
+        _cleanup_(json_variant_unrefp) JsonVariant *v = NULL;
+        _cleanup_free_ void *blob = NULL, *policy_hash = NULL;
+        size_t blob_size = 0, policy_hash_size = 0;
+
+        r = json_parse(json, 0, &v, NULL, NULL);
+        if (r < 0) {
+                crypt_log_dbg(cd, "Could not parse " TOKEN_NAME " json object%s.",
+                              r == -ENOMEM ? " (not enough memory)" : "");
+                return 1;
+        }
+
+        w = json_variant_by_key(v, "tpm2-pcrs");
+        if (!w || !json_variant_is_array(w)) {
+                crypt_log_dbg(cd, "TPM2 token data lacks 'tpm2-pcrs' field.");
+                return 1;
+        }
+
+        JSON_VARIANT_ARRAY_FOREACH(e, w) {
+                uintmax_t u;
+
+                if (!json_variant_is_number(e)) {
+                        crypt_log_dbg(cd, "TPM2 PCR is not a number.");
+                        return 1;
+                }
+
+                u = json_variant_unsigned(e);
+                if (u >= TPM2_PCRS_MAX) {
+                        crypt_log_dbg(cd, "TPM2 PCR number out of range.");
+                        return 1;
+                }
+        }
+
+        w = json_variant_by_key(v, "tpm2-blob");
+        if (!w || !json_variant_is_string(w)) {
+                crypt_log_dbg(cd, "TPM2 token data lacks 'tpm2-blob' field.");
+                return 1;
+        }
+
+        if (unbase64mem(json_variant_string(w), SIZE_MAX, &blob, &blob_size) < 0) {
```

Comment:
give that this function shall only validate, but the data is not used we should probably fix unbase64mem() so that it accepts the final two params as NULL, and then still parses the data, but won't bother returning it.

(i prepped #19734 now, which adds support for this. please update your PR hence, to pass NULL as the final two params here)